File: engine/core/task_system.cpp
#include "task_system.hpp"
#include <algorithm>

void TaskManager::destroy(TaskHandle handle) {
    for (auto& layer : layers_) {
        for (auto& entry : layer) {
            if (entry.id == handle.id) {
                entry.pending_destroy = true;
                return;
            }
        }
    }
}

void TaskManager::run(RunPhase phase, float dt, RenderContext* ctx) {
    for (auto& layer : layers_) {
        switch (phase) {
            case RunPhase::Start:
                for (auto& entry : layer) {
                    if (!entry.initialized && !entry.pending_destroy) {
                        entry.impl->start();
                        entry.initialized = true;
                    }
                }
                break;

            case RunPhase::Update:
                // Initialize any new tasks first (lazy init)
                for (auto& entry : layer) {
                    if (!entry.initialized && !entry.pending_destroy) {
                        entry.impl->start();
                        entry.initialized = true;
                    }
                }
                for (auto& entry : layer) {
                    if (entry.initialized && !entry.pending_destroy) {
                        entry.impl->update(dt);
                    }
                }
                break;

            case RunPhase::Physics:
                for (auto& entry : layer) {
                    if (entry.initialized && !entry.pending_destroy &&
                        entry.impl->has_physics()) {
                        entry.impl->physics(dt);
                    }
                }
                break;

            case RunPhase::Draw:
                if (ctx) {
                    for (auto& entry : layer) {
                        if (entry.initialized && !entry.pending_destroy &&
                            entry.impl->has_draw()) {
                            entry.impl->draw(*ctx);
                        }
                    }
                }
                break;

            case RunPhase::Destroy:
                // Release and remove destroyed tasks
                for (auto& entry : layer) {
                    if (entry.pending_destroy && entry.initialized) {
                        entry.impl->release();
                    }
                }
                layer.erase(
                    std::remove_if(layer.begin(), layer.end(),
                        [](const TaskEntry& e) { return e.pending_destroy; }),
                    layer.end()
                );
                break;
        }
    }
}

size_t TaskManager::task_count() const {
    size_t total = 0;
    for (const auto& layer : layers_) {
        total += layer.size();
    }
    return total;
}

size_t TaskManager::task_count(TaskLayer layer) const {
    auto idx = static_cast<size_t>(layer);
    if (idx >= layers_.size()) return 0;
    return layers_[idx].size();
}

std::vector<TaskManager::TaskThreadingInfo> TaskManager::threading_report() const {
    std::vector<TaskThreadingInfo> report;
    for (size_t li = 0; li < layers_.size(); ++li) {
        auto layer = static_cast<TaskLayer>(li);
        for (const auto& entry : layers_[li]) {
            if (entry.pending_destroy) continue;
            report.push_back({
                entry.id,
                layer,
                entry.impl->threading_policy(),
                entry.impl->is_thread_aware()
            });
        }
    }
    return report;
}

TaskManager::ThreadingSummary TaskManager::threading_summary() const {
    ThreadingSummary summary;
    for (const auto& layer : layers_) {
        for (const auto& entry : layer) {
            if (entry.pending_destroy) continue;
            ++summary.total;
            switch (entry.impl->threading_policy()) {
                case ThreadingPolicy::MainThread: ++summary.main_thread; break;
                case ThreadingPolicy::AnyThread:  ++summary.any_thread;  break;
                case ThreadingPolicy::Parallel:   ++summary.parallel;    break;
            }
        }
    }
    return summary;
}

File: web/src/engine/task_system.ts
import type { TaskLike, Drawable, RenderContext } from './interfaces.js';

export interface TaskHandle {
    id: number;
}

export enum TaskLayer {
    Default = 0,
    Bullet = 1,
    UI = 2,
    Max = 3,
}

export enum RunPhase {
    Start,
    Update,
    Physics,
    Draw,
    Destroy,
}

interface TaskEntry {
    id: number;
    task: TaskLike & Partial<Drawable>;
    initialized: boolean;
    pendingDestroy: boolean;
}

export class TaskManager {
    private layers: TaskEntry[][] = [];
    private nextId = 1;

    constructor() {
        for (let i = 0; i < TaskLayer.Max; i++) {
            this.layers.push([]);
        }
    }

    registerTask(task: TaskLike & Partial<Drawable>, layer: TaskLayer = TaskLayer.Default): TaskHandle {
        const id = this.nextId++;
        this.layers[layer].push({ id, task, initialized: false, pendingDestroy: false });
        return { id };
    }

    destroy(handle: TaskHandle): void {
        for (const layer of this.layers) {
            for (const entry of layer) {
                if (entry.id === handle.id) {
                    entry.pendingDestroy = true;
                    return;
                }
            }
        }
    }

    run(phase: RunPhase, dt: number, ctx?: RenderContext): void {
        for (const layer of this.layers) {
            switch (phase) {
                case RunPhase.Start:
                    for (const entry of layer) {
                        if (!entry.initialized && !entry.pendingDestroy) {
                            entry.task.start();
                            entry.initialized = true;
                        }
                    }
                    break;

                case RunPhase.Update:
                    for (const entry of layer) {
                        if (!entry.initialized && !entry.pendingDestroy) {
                            entry.task.start();
                            entry.initialized = true;
                        }
                    }
                    for (const entry of layer) {
                        if (entry.initialized && !entry.pendingDestroy) {
                            entry.task.update(dt);
                        }
                    }
                    break;

                case RunPhase.Draw:
                    if (ctx) {
                        for (const entry of layer) {
                            if (entry.initialized && !entry.pendingDestroy && entry.task.draw) {
                                entry.task.draw(ctx);
                            }
                        }
                    }
                    break;

                case RunPhase.Destroy:
                    for (let i = layer.length - 1; i >= 0; i--) {
                        if (layer[i].pendingDestroy) {
                            if (layer[i].initialized) {
                                layer[i].task.release();
                            }
                            layer.splice(i, 1);
                        }
                    }
                    break;
            }
        }
    }
}
